flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/translators/SinkTransformationTranslator.java

```diff
@@ -19,250 +19,237 @@
+            if (sink instanceof WithPostCommitTopology) {
+                DataStream<CommittableMessage<CommT>> postcommitted = addFailOverRegion(committed);
+                adjustTransformations(
+                        postcommitted,
+                        pc -> {
+                            ((WithPostCommitTopology<T, CommT>) sink).addPostCommitTopology(pc);
+                            return null;
+                        });
+            }
+        }
 
-        streamGraph.setParallelism(transformationId, parallelism);
-        streamGraph.setMaxParallelism(transformationId, maxParallelism);
+        /**
+         * Adds a batch exchange that materializes the output first. This is a no-op in STREAMING.
+         */
+        private <T> DataStream<T> addFailOverRegion(DataStream<T> input) {
+            return new DataStream<>(
+                    executionEnvironment,
+                    new PartitionTransformation<>(
+                            input.getTransformation(),
+                            new ForwardPartitioner<>(),
+                            StreamExchangeMode.BATCH));
+        }
 
-        StreamGraphUtils.configureBufferTimeout(
-                streamGraph,
-                transformationId,
-                sinkTransformation,
-                context.getDefaultBufferTimeout());
-        if (uid != null) {
-            streamGraph.setTransformationUID(transformationId, uid);
+        private <T, R> R adjustTransformations(
+                DataStream<T> inputStream, Function<DataStream<T>, R> action) {
+            int numTransformsBefore = executionEnvironment.getTransformations().size();
+            R result = action.apply(inputStream);
+            List<Transformation<?>> transformations = executionEnvironment.getTransformations();
+            List<Transformation<?>> expandedTransformations =
+                    transformations.subList(numTransformsBefore, transformations.size());
+            for (Transformation<?> subTransformation : expandedTransformations) {
+                concatUid(
+                        subTransformation,
+                        Transformation::getUid,
+                        Transformation::setUid,
+                        subTransformation.getName());
+                concatProperty(subTransformation, Transformation::getName, Transformation::setName);
+                concatProperty(
+                        subTransformation,
+                        Transformation::getDescription,
+                        Transformation::setDescription);
+
+                Optional<SlotSharingGroup> ssg = transformation.getSlotSharingGroup();
+                if (ssg.isPresent() && !subTransformation.getSlotSharingGroup().isPresent()) {
+                    subTransformation.setSlotSharingGroup(ssg.get());
+                }
+                subTransformation.setParallelism(transformation.getParallelism());
```

Comment:
Is there a good way to determine that the user explicitly set the parallelism? So far I assumed that if the user sets the parallelism of the `SinkTransformation` it applies to all transformations that are expanded.

I also do not see a good a way to distinguish between "user set" parallelism because afaik when a `Transformation` is created you always have to pass a parallelism.

In end, we probably have to define our own interpretation of what setting the parallelism at the `SinkTransformation` implies.

Reply:
It seems we indeed do not have a good method to check if parallelism is set. Logically we should have the priority with `customized > sink transformations' settings > env default`. We might have to support it since for compaction we need to set the compact coordinator to parallelism 1...

One possible solution is still to use a separate environment with the default parallelism set to sink transformation's settings...

Reply:
Hi Fabian~I'm a bit afraid there might be bad cases for this method since users might
```
env.setParallelism(1);
DataStream x = ...
x.addSink(new FinkSink()).setParallism(2)
```

while inside the precommit topology, operators like CompactorCoordiantor is set to 1 specially.

In this case, the final parallelism for the compactor coordinator would still be 2 and break the priority?

Reply:
Hopefully, I found a final solution now I had to change how the parallelism is passed to the transformations 06eb49a. Now defaulting to the env parallelism is the responsibility of the translators and by default, the parallelism is always `ExecutionConfig#DEFAULT_PARALLELISM`. This allows to do:

```java
                if (subTransformation.getParallelism() == ExecutionConfig.PARALLELISM_DEFAULT) {
                    if (transformation.getParallelism() == ExecutionConfig.PARALLELISM_DEFAULT) {
                        subTransformation.setParallelism(
                                context.getStreamGraph().getExecutionConfig().getParallelism());
                    } else {
                        subTransformation.setParallelism(transformation.getParallelism());
                    }
                }

```